### ISSUE TITLE
False positive error on `List` `+` operator when either LHS or RHS's type is a `List` of a `Union`:

### ISSUE THREAD

**Issue by kennipj** [microsoft/pyright#2146]
False positive error on `List` `+` operator when either left hand side or right hand side's type is a `List` of a `Union`:

Examples:

LHS is `Union`:
```py
from typing import List, Union


def some_func(some_list: List[Union[int, str]]):
    some_list + [5]
    # Operator "+" not supported for types "List[int | str]" and "list[int]"
    some_list += [5]  # No error

```

RHS is compatible `Union`:
```py
from typing import ClassVar, List, Protocol


class SomeProtocol(Protocol):
    some_field: ClassVar[int]


def other_func(some_list: List[SomeProtocol], some_obj: SomeProtocol):
    if isinstance(some_obj, type):
        ...
    some_list + [some_obj]
    # Operator "+" not supported for types "List[SomeProtocol]" and
    #  "list[<subclass of SomeProtocol and type> | SomeProtocol]"
    some_list += [some_obj]  # No error

```

Tested on pyright version 1.1.159

**Comment by kennipj** [comment]
One thing I notice is that RHS shows list with lowercase ie. `list[int]`, as opposed to `List[int]`. Maybe that's the culprit?

**Comment by erictraut** [comment]
This is due to a complex interaction between binary operator logic and bidirectional type inference. Binary operators in Python are quite complicated because they are implemented with one of several magic methods. For example, the `+` operator is implemented with `__add__` and `__radd__`. If the two operands are simple (non-union) types, it's relatively straightforward to determine whether the `__add__` or `__radd__` method applies. Once unions are involved, it gets much more complicated because all combinations of union subtypes need to be considered. Some combinations may use `__add__` and other combinations might use `__radd__`. For this reason, pyright evaluates the types of the LHS and RHS prior to determining the type of the binary expression. When evaluating the types of both operands upfront, bidirectional type inference cannot be applied. That means we fall back on normal inference rules. In this case, that means `[5]` is evaluated as type `list[int]`, and that cannot be added to type `list[int | str]`.

The `+=` operator works OK in this case because the `list` class (or rather the `MutableSequence` class, from which `list` derives) provides an `__iadd__` method. In this case, the RHS operand can be evaluated with bidirectional type inference. So the analyzer is able to infer that the type of `[5]` is `list[int | str]` because it has additional context.

I'll think a bit more about ways to make this better. So far, the only ways I can think of would have a significant and measurable performance impact on the type analysis of all binary operators. Since these are so common in python code, I don't think it would be a good tradeoff to handle this specific edge case.

The workaround is to use an explicit type annotation for your list expression.

```python
val: List[Union[int, str]] = [5]
some_list + val
```

**Comment by erictraut** [comment]
I thought of a cheap heuristic that will handle the specific case for the "+" operator where the RHS operand is a list expression. It's a bit of a hack in that it's specific to this one case, but I think it's justified because this is such a common pattern in Python. This change will be in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.161, which I just published. It will also be included in the next release of pylance.
